File: 10a.pl
#!/usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;


{   package SyntaxChecker;
    use Moo;
    has _stack => (is      => 'lazy',
                   builder => sub { [] },
                   clearer => '_clean_stack');

    my %score = (
        ')' => 3,
        ']' => 57,
        '}' => 1197,
        '>' => 25137,
    );
    my %expect = qw/ ( ) [ ] { } < > /;

    sub push_stack {
        my ($self, $ch) = @_;
        push @{ $self->_stack }, $ch;
    }

    sub pop_stack {
        my ($self) = @_;
        return pop @{ $self->_stack }
    }

    sub score {
        my ($self, $line) = @_;
        $self->_clean_stack;
        for my $ch (split //, $line) {
            if (exists $expect{$ch}) {
                $self->push_stack($expect{$ch});

            } elsif ($ch ne $self->pop_stack) {
                return $score{$ch}
            }
        }
        return 0
    }
}

my $score = 0;
my $sc = 'SyntaxChecker'->new;
while (<>) {
    chomp;
    $score += $sc->score($_);
}
say $score;

__DATA__
[({(<(())[]>[[{[]{<()<>>
[(()[<>])]({[<{<<[]>>(
{([(<{}[<>[]}>{[]{[(<()>
(((({<>}<{<{<>}{[]{[]{}
[[<[([]))<([[{}[[()]]]
[{[{({}]{}}([{[{{{}}([]
{<[[]]>}<{[{[{[]{()[[[]
[<(<(<(<{}))><([]([]()
<{([([[(<>()){}]>(<<{{
<{([{{}}[<[[[<>{}]]]>[]]

File: 10a.py
#! /usr/bin/python3
import sys

class SyntaxChecker:
    stack = []

    scores = {')': 3, ']': 57, '}': 1197, '>': 25137}
    expect = {'(':')', '[':']', '{':'}', '<':'>'}

    def clean_stack (self):
        self.stack = []

    def push_stack (self, ch):
        self.stack.append(ch)

    def pop_stack (self):
        return self.stack.pop()

    def score (self, line):
        self.clean_stack()
        for ch in line:
            if ch in self.expect:
                self.push_stack(self.expect[ch])
            elif ch != self.pop_stack():
                return self.scores[ch]
        return 0

score = 0
sc = SyntaxChecker()

with open(sys.argv[1], 'r') as fh:
    for line in fh:
        line = line[:-1]
        score += sc.score(line)
print(score)
